File: quadratic_solver.go
// quadratic_solver.go - Решатель квадратных уравнений на Go (CLI)
package main

import (
	"bufio"
	"flag"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type Result struct {
	Type   string  // "real", "double", "complex", "linear", "none"
	D      float64
	X1     float64
	X2     float64
	Real   float64
	Imag   float64
	Root   float64
}

func solveQuadratic(a, b, c float64) Result {
	if a == 0 {
		if b == 0 {
			return Result{Type: "none"}
		}
		return Result{Type: "linear", Root: -c / b}
	}
	D := b*b - 4*a*c
	if D > 0 {
		sqrtD := math.Sqrt(D)
		return Result{Type: "real", D: D, X1: (-b + sqrtD) / (2 * a), X2: (-b - sqrtD) / (2 * a)}
	} else if D == 0 {
		return Result{Type: "double", D: D, X1: -b / (2 * a)}
	} else {
		real := -b / (2 * a)
		imag := math.Sqrt(-D) / (2 * a)
		return Result{Type: "complex", D: D, Real: real, Imag: imag}
	}
}

func formatComplex(real, imag float64) string {
	if imag >= 0 {
		return fmt.Sprintf("%.4f + %.4fi", real, imag)
	} else {
		return fmt.Sprintf("%.4f - %.4fi", real, -imag)
	}
}

func main() {
	var aFlag, bFlag, cFlag float64
	flag.Float64Var(&aFlag, "a", 0, "Коэффициент a")
	flag.Float64Var(&bFlag, "b", 0, "Коэффициент b")
	flag.Float64Var(&cFlag, "c", 0, "Коэффициент c")
	flag.Parse()

	var a, b, c float64
	if flag.NFlag() == 3 {
		a, b, c = aFlag, bFlag, cFlag
	} else {
		reader := bufio.NewReader(os.Stdin)
		fmt.Print("Введите коэффициент a: ")
		input, _ := reader.ReadString('\n')
		a, _ = strconv.ParseFloat(strings.TrimSpace(input), 64)
		fmt.Print("Введите коэффициент b: ")
		input, _ = reader.ReadString('\n')
		b, _ = strconv.ParseFloat(strings.TrimSpace(input), 64)
		fmt.Print("Введите коэффициент c: ")
		input, _ = reader.ReadString('\n')
		c, _ = strconv.ParseFloat(strings.TrimSpace(input), 64)
	}

	fmt.Printf("\nУравнение: %.1fx² + %.1fx + %.1f = 0\n", a, b, c)
	res := solveQuadratic(a, b, c)

	switch res.Type {
	case "none":
		if c == 0 {
			fmt.Println("Бесконечное множество решений (0 = 0).")
		} else {
			fmt.Println("Нет решений (противоречие).")
		}
	case "linear":
		fmt.Printf("Линейное уравнение, корень: %.4f\n", res.Root)
	case "real":
		fmt.Printf("Дискриминант D = %.4f\n", res.D)
		fmt.Printf("Корни:\nx₁ = %.4f\nx₂ = %.4f\n", res.X1, res.X2)
	case "double":
		fmt.Printf("Дискриминант D = %.4f\n", res.D)
		fmt.Printf("Корень (двойной): x = %.4f\n", res.X1)
	case "complex":
		fmt.Printf("Дискриминант D = %.4f\n", res.D)
		fmt.Println("Комплексные корни:")
		fmt.Printf("x₁ = %s\n", formatComplex(res.Real, res.Imag))
		fmt.Printf("x₂ = %s\n", formatComplex(res.Real, -res.Imag))
	}
}

File: quadratic_solver.rs
// quadratic_solver.rs - Решатель квадратных уравнений на Rust (CLI)
use std::io;
use std::io::Write;
use std::f64;

#[derive(Debug)]
enum Solution {
    None,
    Linear(f64),
    Real { d: f64, x1: f64, x2: f64 },
    Double { d: f64, x: f64 },
    Complex { d: f64, real: f64, imag: f64 },
}

fn solve_quadratic(a: f64, b: f64, c: f64) -> Solution {
    if a == 0.0 {
        if b == 0.0 {
            return Solution::None;
        }
        return Solution::Linear(-c / b);
    }
    let d = b * b - 4.0 * a * c;
    if d > 0.0 {
        let sqrt_d = d.sqrt();
        let x1 = (-b + sqrt_d) / (2.0 * a);
        let x2 = (-b - sqrt_d) / (2.0 * a);
        Solution::Real { d, x1, x2 }
    } else if d == 0.0 {
        let x = -b / (2.0 * a);
        Solution::Double { d, x }
    } else {
        let real = -b / (2.0 * a);
        let imag = (-d).sqrt() / (2.0 * a);
        Solution::Complex { d, real, imag }
    }
}

fn format_complex(real: f64, imag: f64) -> String {
    if imag >= 0.0 {
        format!("{:.4} + {:.4}i", real, imag)
    } else {
        format!("{:.4} - {:.4}i", real, -imag)
    }
}

fn read_float(prompt: &str) -> f64 {
    print!("{}", prompt);
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Ошибка ввода");
    input.trim().parse().expect("Введите число")
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let (a, b, c) = if args.len() == 4 {
        let a = args[1].parse().expect("Неверный a");
        let b = args[2].parse().expect("Неверный b");
        let c = args[3].parse().expect("Неверный c");
        (a, b, c)
    } else {
        println!("Решатель квадратных уравнений");
        let a = read_float("Введите коэффициент a: ");
        let b = read_float("Введите коэффициент b: ");
        let c = read_float("Введите коэффициент c: ");
        (a, b, c)
    };

    println!("\nУравнение: {:.1}x² + {:.1}x + {:.1} = 0", a, b, c);
    let solution = solve_quadratic(a, b, c);

    match solution {
        Solution::None => {
            if c == 0.0 {
                println!("Бесконечное множество решений (0 = 0).");
            } else {
                println!("Нет решений (противоречие).");
            }
        }
        Solution::Linear(root) => {
            println!("Линейное уравнение, корень: {:.4}", root);
        }
        Solution::Real { d, x1, x2 } => {
            println!("Дискриминант D = {:.4}", d);
            println!("Корни:\nx₁ = {:.4}\nx₂ = {:.4}", x1, x2);
        }
        Solution::Double { d, x } => {
            println!("Дискриминант D = {:.4}", d);
            println!("Корень (двойной): x = {:.4}", x);
        }
        Solution::Complex { d, real, imag } => {
            println!("Дискриминант D = {:.4}", d);
            println!("Комплексные корни:");
            println!("x₁ = {}", format_complex(real, imag));
            println!("x₂ = {}", format_complex(real, -imag));
        }
    }
}
